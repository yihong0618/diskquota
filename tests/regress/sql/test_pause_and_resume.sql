-- Test pause and resume.
CREATE SCHEMA s1;
SET search_path TO s1;

CREATE TABLE a(i int) DISTRIBUTED BY (i);

-- expect insert succeed
INSERT INTO a SELECT generate_series(1,100000);

SELECT diskquota.set_schema_quota('s1', '1 MB');
SELECT diskquota.wait_for_worker_new_epoch();
-- expect insert fail
INSERT INTO a SELECT generate_series(1,100);

-- pause extension
SELECT diskquota.pause();
SELECT diskquota.wait_for_worker_new_epoch();

SELECT tableid::regclass, size, segid FROM diskquota.table_size 
WHERE tableid = 'a'::regclass AND segid = -1;

-- expect insert succeed
INSERT INTO a SELECT generate_series(1,100000);

-- resume extension
SELECT diskquota.resume();
SELECT diskquota.wait_for_worker_new_epoch();

-- expect insert fail
INSERT INTO a SELECT generate_series(1,100);

-- table size should be updated after resume
SELECT tableid::regclass, size, segid FROM diskquota.table_size 
WHERE tableid = 'a'::regclass AND segid = -1;

-- enable hardlimit
\! gpconfig -c "diskquota.hard_limit" -v "on" > /dev/null
\! gpstop -u
SELECT diskquota.wait_for_worker_new_epoch();

-- blackmaps should be dispatched to segments
SELECT count(*) FROM (select diskquota.show_blackmap() from gp_dist_random('gp_id')) as blackmap_on_segs;

-- expect insert fail
INSERT INTO a SELECT generate_series(1,100);

-- pause extension
SELECT diskquota.pause();
SELECT diskquota.wait_for_worker_new_epoch();

-- pause should clear blackmap on segments
SELECT count(*) FROM (select diskquota.show_blackmap() from gp_dist_random('gp_id')) as blackmap_on_segs;
-- expect insert succeed
INSERT INTO a SELECT generate_series(1,100);

-- resume extension
SELECT diskquota.pause();
SELECT diskquota.wait_for_worker_new_epoch();

-- blackmaps should be dispatched to segments again
SELECT count(*) FROM (select diskquota.show_blackmap() from gp_dist_random('gp_id')) as blackmap_on_segs;

-- expect insert fail
INSERT INTO a SELECT generate_series(1,100);


-- disable hardlimit
\! gpconfig -c "diskquota.hard_limit" -v "off" > /dev/null
\! gpstop -u

RESET search_path;
DROP TABLE s1.a;
DROP SCHEMA s1;
