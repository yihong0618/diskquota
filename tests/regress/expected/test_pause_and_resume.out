-- Test pause and resume.
CREATE SCHEMA s1;
SET search_path TO s1;
CREATE TABLE a(i int) DISTRIBUTED BY (i);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'i' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
-- expect insert succeed
INSERT INTO a SELECT generate_series(1,100000);
SELECT diskquota.set_schema_quota('s1', '1 MB');
 set_schema_quota 
------------------
 
(1 row)

SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t
(1 row)

-- expect insert fail
INSERT INTO a SELECT generate_series(1,100);
ERROR:  schema's disk space quota exceeded with name: s1
-- pause extension
SELECT diskquota.pause();
 pause 
-------
 
(1 row)

SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t
(1 row)

SELECT tableid::regclass, size, segid FROM diskquota.table_size 
WHERE tableid = 'a'::regclass AND segid = -1;
 tableid |  size   | segid 
---------+---------+-------
 a       | 3932160 |    -1
(1 row)

-- expect insert succeed
INSERT INTO a SELECT generate_series(1,100000);
-- resume extension
SELECT diskquota.resume();
 resume 
--------
 
(1 row)

SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t
(1 row)

-- expect insert fail
INSERT INTO a SELECT generate_series(1,100);
ERROR:  schema's disk space quota exceeded with name: s1
-- table size should be updated after resume
SELECT tableid::regclass, size, segid FROM diskquota.table_size 
WHERE tableid = 'a'::regclass AND segid = -1;
 tableid |  size   | segid 
---------+---------+-------
 a       | 7569408 |    -1
(1 row)

RESET search_path;
DROP TABLE s1.a;
DROP SCHEMA s1;
